File: algorithms/coding-interview-questions-blind75-lol/3/insert_interval.py
# ugly
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        if not intervals:
            return [newInterval]

        added = False
        ret = []
        for s, e in intervals:
            if e < newInterval[0] or (s > newInterval[1] and added):
                ret.append([s, e])
            else:
                added = True
                if len(ret) and ret[-1][1] >= newInterval[0]:
                    ret[-1][1] = max(newInterval[1], e)
                else:
                    if s > newInterval[1]:
                        ret.append(newInterval)
                        ret.append([s, e])
                    else:
                        new_i, new_j = min(newInterval[0], s), max(newInterval[1], e)
                        ret.append([new_i, new_j])

        if not added:
            ret.append(newInterval)

        return ret

File: algorithms/coding-interview-questions-blind75-lol/3/insert_interval.cpp
// clean solution

class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        vector<vector<int>> ans;

        int s = newInterval[0], e = newInterval[1];

        int i = 0, n = size(intervals);

        while (i < n && intervals[i][1] < s)
            ans.push_back(intervals[i++]);

        while (i < n && e >= intervals[i][0]) {
            s = min(s, intervals[i][0]);
            e = max(e, intervals[i][1]);
            ++i;
        }

        ans.push_back({s, e});

        while (i < n)
            ans.push_back(intervals[i++]);
        
        return ans;
    }
};
